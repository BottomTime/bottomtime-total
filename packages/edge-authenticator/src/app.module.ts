import { Module } from '@nestjs/common';
import { PassportModule } from '@nestjs/passport';

import { AppController } from './app.controller';
import { DependenciesModule } from './dependencies.module';
import { GoogleStrategy } from './google/google.strategy';
import { JwtService } from './jwt/jwt.service';
import { JwtStrategy } from './jwt/jwt.strategy';
import { UserService } from './user.service';

@Module({
  imports: [
    PassportModule.register({
      session: false,
    }),
    DependenciesModule.forRoot(),
  ],
  providers: [UserService, JwtService, JwtStrategy, GoogleStrategy],
  controllers: [AppController],
})
export class AppModule {}
