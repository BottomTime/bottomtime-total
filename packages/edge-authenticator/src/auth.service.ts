import { Injectable, UnauthorizedException } from '@nestjs/common';

import { JwtPayload } from 'jsonwebtoken';

import { User } from './user';

@Injectable()
export class AuthService {
  async validateJwt(payload: JwtPayload): Promise<User> {
    if (!payload.sub) {
      throw new UnauthorizedException('JWT payload did not contain a subject.');
    }

    return { email: payload.sub };
  }
}
