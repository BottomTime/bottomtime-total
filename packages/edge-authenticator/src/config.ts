/* eslint-disable no-process-env */
import 'dotenv/config';
import { z } from 'zod';

const LogLevelSchema = z.enum([
  'trace',
  'debug',
  'info',
  'warn',
  'error',
  'fatal',
]);
type LogLevel = z.infer<typeof LogLevelSchema>;

export interface AppConfig {
  awsRegion: string;
  baseUrl: string;
  cognito: {
    clientId: string;
    clientSecret: string;
    domain: string;
  };
  cookieName: string;
  isProduction: boolean;
  env: string;
  logLevel: LogLevel;
  sessionSecret: string;
}

const ConfigSchema = z
  .object({
    AWS_REGION: z.string().default('us-east-1'),
    BT_EDGE_BASE_URL: z.string().default('http://localhost:9000'),
    BT_EDGE_CLIENT_ID: z.string().default(''),
    BT_EDGE_CLIENT_SECRET: z.string().default(''),
    BT_EDGE_COGNITO_DOMAIN: z.string().default(''),
    BT_EDGE_COOKIE_NAME: z.string().default('bottomtime.auth'),
    BT_EDGE_SESSION_SECRET: z
      .string()
      .default(
        'nxS0JJ04kNjiZpJxQz5iq6OFoN6bAvsQxO2eVLGaSQyslZU8ltxqYlmKUIon9B8scg89VBg3eFZAs6umkWUYWQ',
      ),
    BT_LOG_LEVEL: LogLevelSchema.default('debug'),
    NODE_ENV: z.string().default('local'),
  })
  .transform<AppConfig>((env) => ({
    awsRegion: env.AWS_REGION,
    baseUrl: env.BT_EDGE_BASE_URL,
    cognito: {
      domain: env.BT_EDGE_COGNITO_DOMAIN,
      clientId: env.BT_EDGE_CLIENT_ID,
      clientSecret: env.BT_EDGE_CLIENT_SECRET,
    },
    cookieName: env.BT_EDGE_COOKIE_NAME,
    isProduction: env.NODE_ENV === 'production',
    env: env.NODE_ENV,
    logLevel: env.BT_LOG_LEVEL,
    sessionSecret: env.BT_EDGE_SESSION_SECRET,
  }));

export const Config = ConfigSchema.parse(process.env);
