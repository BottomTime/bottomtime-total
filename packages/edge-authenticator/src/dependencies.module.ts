import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamicModule, Module } from '@nestjs/common';

@Module({})
export class DependenciesModule {
  static forRoot(): DynamicModule {
    return {
      module: DependenciesModule,
      providers: [
        {
          provide: DynamoDBClient,
          useFactory: () => new DynamoDBClient(),
        },
      ],
      exports: [DynamoDBClient],
    };
  }
}
