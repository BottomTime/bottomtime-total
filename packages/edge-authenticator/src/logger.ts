import { LoggerService } from '@nestjs/common';

import Logger from 'bunyan';
import { z } from 'zod';

const LogLevel = z.enum(['trace', 'debug', 'info', 'warn', 'error', 'fatal']);
type LogLevel = z.infer<typeof LogLevel>;

export function createLogger(logLevel?: string): Logger {
  let level: LogLevel = 'info';

  const parsed = LogLevel.safeParse(logLevel);
  if (parsed.success) {
    level = parsed.data;
  }

  const logger = Logger.createLogger({
    name: 'bottomtime',
    level,
    stream: process.stdout,
    serializers: { err: Logger.stdSerializers.err },
  });

  return logger;
}

export class BunyanLoggerService implements LoggerService {
  constructor(private readonly logger: Logger) {}

  log(message: unknown, ...optionalParams: unknown[]) {
    this.logger.info(message, ...optionalParams);
  }

  error(message: unknown, ...optionalParams: unknown[]) {
    this.logger.error(message, ...optionalParams);
  }

  warn(message: unknown, ...optionalParams: unknown[]) {
    this.logger.warn(message, ...optionalParams);
  }

  debug(message: unknown, ...optionalParams: unknown[]) {
    this.logger.debug(message, ...optionalParams);
  }

  verbose(message: unknown, ...optionalParams: unknown[]) {
    this.logger.trace(message, ...optionalParams);
  }

  fatal(message: unknown, ...optionalParams: unknown[]) {
    this.logger.fatal(message, ...optionalParams);
  }
}
