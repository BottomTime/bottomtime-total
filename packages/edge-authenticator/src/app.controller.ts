import {
  Controller,
  Get,
  Inject,
  Logger,
  Redirect,
  Render,
  Res,
  UseGuards,
} from '@nestjs/common';

import { Response } from 'express';

import { Config } from './config';
import { CurrentUser } from './current-user.decorator';
import { GoogleAuthGuard } from './google/google.guard';
import { JwtAuthGuard } from './jwt/jwt.guard';
import { JwtService } from './jwt/jwt.service';
import { User } from './user';

@Controller('/')
@UseGuards(JwtAuthGuard)
export class AppController {
  private readonly log = new Logger(AppController.name);

  constructor(@Inject(JwtService) private readonly jwtService: JwtService) {}

  @Get()
  @Render('index')
  async index(@CurrentUser() user: User | undefined) {
    this.log.debug(
      `Received request for home page from <${user?.email || 'anonymous'}>.`,
    );
    return { user: user?.email, domain: Config.protectedDomain };
  }

  @Get('login')
  @UseGuards(GoogleAuthGuard)
  login() {
    /* Nothing to do here. Passport will redirect to AWS Cognito client. */
    this.log.debug('Redirecting to Google for authentication.');
  }

  @Get('callback')
  @UseGuards(GoogleAuthGuard)
  @Render('authorized')
  callback(@CurrentUser() user: User, @Res() res: Response) {
    this.log.debug('Received callback from Google.');

    const jwt = this.jwtService.signJwt(user.email, user.authorizedDomains);

    this.log.debug(
      `Generated JWT for <${user.email}>. Issuing cookie: ${Config.cookie.name} for domain ${Config.cookie.domain}.`,
    );
    res.cookie(Config.cookie.name, jwt, {
      domain: Config.cookie.domain,
      httpOnly: true,
      maxAge: Config.cookie.ttl * 1000,
    });

    return { user: user.email, domain: Config.protectedDomain };
  }

  @Get('logout')
  @Redirect('/')
  logout(@Res() res: Response) {
    res.clearCookie(Config.cookie.name);
  }
}
