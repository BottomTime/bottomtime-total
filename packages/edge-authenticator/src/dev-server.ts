import 'dotenv/config';

import { createApp } from './app';
import { Config } from './config';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then((app) => {
    app.listen(9000);
    log.info('ðŸš€ Authenticator server running at http://localhost:9000');
  })
  .catch((error) => {
    log.fatal(error);
    process.exit(1);
  });
