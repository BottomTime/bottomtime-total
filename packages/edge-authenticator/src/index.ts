import serverless from '@codegenie/serverless-express';

import { Callback, Context, Handler } from 'aws-lambda';

import { createApp } from './app';
import { Config } from './config';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);
let cachedServer: Handler;

async function init(): Promise<void> {
  if (!cachedServer) {
    log.info('ðŸš€ Initializing Edge Authorizer...');
    const app = await createApp(log);
    await app.init();
    cachedServer = serverless({ app: app.getHttpAdapter().getInstance() });
    log.info('ðŸŽ‰ Edge Authorizer is ready! ðŸŽ‰');
  }
}

export const handler: Handler = async (
  event: unknown,
  context: Context,
  cb: Callback,
): Promise<unknown> => {
  try {
    await init();
    return await cachedServer(event, context, cb);
  } catch (error) {
    log.fatal(error);
    return { isAuthorized: false };
  }
};
