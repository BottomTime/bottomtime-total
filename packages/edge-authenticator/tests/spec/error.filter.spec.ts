import {
  ArgumentsHost,
  BadRequestException,
  HttpStatus,
  UnauthorizedException,
} from '@nestjs/common';

import { Response } from 'express';
import { createRequest, createResponse } from 'node-mocks-http';

import { GlobalErrorFilter } from '../../src/error.filter';
import { BunyanLoggerService } from '../../src/logger';
import { Log } from '../logger';

function getArgumentHost(res: Response): ArgumentsHost {
  return {
    switchToHttp: () => ({
      getRequest: () =>
        createRequest({
          method: 'GET',
          path: '/path/to/resource',
        }),
      getResponse: () => res,
    }),
  } as ArgumentsHost;
}

describe('Global error filter', () => {
  let logService: BunyanLoggerService;
  let filter: GlobalErrorFilter;

  beforeAll(() => {
    logService = new BunyanLoggerService(Log);
    filter = new GlobalErrorFilter(logService);
  });

  it('will return a basic 500 error for an unknown exception', () => {
    const res = createResponse();
    filter.catch('An unknown error', getArgumentHost(res));
    expect(res._isEndCalled()).toBe(true);
    expect(res._getStatusCode()).toBe(500);
    expect(res._getJSONData()).toMatchSnapshot();
  });

  it('will return more details if the exception is an Error object', () => {
    const error = new Error('An error object');
    const res = createResponse();

    filter.catch(error, getArgumentHost(res));

    expect(res._isEndCalled()).toBe(true);
    expect(res._getStatusCode()).toBe(500);
    expect(res._getJSONData()).toMatchSnapshot();
  });

  it('will return even more details if the exception is an HttpException', () => {
    const error = new BadRequestException('This is a bad request exception');
    const res = createResponse();

    filter.catch(error, getArgumentHost(res));

    expect(res._isEndCalled()).toBe(true);
    expect(res._getStatusCode()).toBe(HttpStatus.BAD_REQUEST);
    expect(res._getJSONData()).toMatchSnapshot();
  });

  it('will attempt to invalidate the session cookie and render error page if the exception is an UNAUTHORIZED HttpException', () => {
    const error = new UnauthorizedException('Nope.');
    const res = createResponse();

    filter.catch(error, getArgumentHost(res));

    expect(res._isEndCalled()).toBe(true);
    expect(res._getStatusCode()).toBe(HttpStatus.UNAUTHORIZED);
    expect(res._getRenderView()).toBe('unauthorized');

    // expect(res._getHeaders()).toBe('');
  });
});
