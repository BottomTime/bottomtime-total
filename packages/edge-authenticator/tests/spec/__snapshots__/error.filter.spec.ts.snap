// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global error filter will return a basic 500 error for an unknown exception 1`] = `
{
  "message": "Internal server error",
  "method": "GET",
  "path": "/path/to/resource",
  "status": 500,
}
`;

exports[`Global error filter will return even more details if the exception is an HttpException 1`] = `
{
  "message": "This is a bad request exception",
  "method": "GET",
  "path": "/path/to/resource",
  "status": 400,
}
`;

exports[`Global error filter will return more details if the exception is an Error object 1`] = `
{
  "message": "An error object",
  "method": "GET",
  "path": "/path/to/resource",
  "status": 500,
}
`;
